new(driverkit): allow configurable builder images

LucaGuerra opened the pull request:
Signed-off-by: Luca Guerra <luca@guerra.sh>

This PR is co-authored with @Issif .

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/CONTRIBUTING.md) file.
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area build

> /area cmd

> /area pkg

> /area docs

> /area tests


**What this PR does / why we need it**:

The underlying problem that we're trying to solve is that some new kernel versions might require specific versions of base system libraries (mostly GNU libc) and/or specific compilers in order to build headers or modules compatible with that kernel. Think about GLIBC requirements for Ubuntu Impish or the latest Fedora.

A good step in doing that is allowing the builder image to be configurable. This allows the user to quickly test and configure which image to use to see if it fixes specific issues with a specific distribution. Due to how Driverkit is designed, this can be specified both on the CLI or in the configuration files.

Thanks to @Issif we have improved this PR in various ways, including making it clear which image is used, even when getting help on the command line.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
new(driverkit): add builderimage option to specify a custom image to be used for building
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/driverkit/pull/113#" title="Author self-approved">LucaGuerra</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fntlnz** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fntlnz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fdriverkit).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/driverkit/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fntlnz"]} -->

Issif commented:
I like the idea, we saw that for newer kernels for debian we had issues with current image and we need to update it. Allowing us, and users, to create and use their own image will be useful. The point I don't know is how to map an image with an os/kernel version.

LucaGuerra commented on pkg/driverbuilder/buildprocessor.go:
```diff
@@ -8,7 +8,7 @@ type BuildArchitecture string
 const BuildArchitectureX86_64 BuildArchitecture = "x86_64"
 
-var builderBaseImage = "falcosecurity/driverkit-builder:latest" // This is overwritten when using the Makefile to build
+var BuilderBaseImage = "falcosecurity/driverkit-builder:latest" // This is overwritten when using the Makefile to build
```
If `BuilderBaseImage` changes name, do we need to reflect this change in the makefile/build process too?

Issif replied:
You're right, I forgot.

LucaGuerra commented on cmd/root.go:
```diff
@@ -117,7 +118,7 @@ func NewRootCmd() *RootCmd {
 	flags.StringVar(&rootOpts.ModuleDeviceName, "moduledevicename", rootOpts.ModuleDeviceName, "kernel module device name (the default is falco, so the device will be under /dev/falco*)")
 	flags.StringVar(&rootOpts.ModuleDriverName, "moduledrivername", rootOpts.ModuleDriverName, "kernel module driver name, i.e. the name you see when you check installed modules via lsmod")
-	flags.StringVar(&rootOpts.BuilderImage, "builderimage", rootOpts.BuilderImage, "docker image to be used to build the kernel module and eBPF probe. If not provided, the default image will be used.")
+	flags.StringVar(&rootOpts.BuilderImage, "builderimage", driverbuilder.BuilderBaseImage, "docker image to be used to build the kernel module and eBPF probe. If not provided, the default image will be used.")
```
If this changes the `BuilderBaseImage` directly the previous change that creates a `CustomBaseImage` clashes a bit with that. Which do you think it's better so we can keep only one?

Issif replied:
We saw together, I'm fixing that too.

leogr commented:
Hey @leodido / @fntlnz 

Is anyone up for a review? :pray:

FedeDP commented:
I think this one can really help to fix #143 issues with ubuntu and debian, as we might need to user newer gcc versions for those (and possibly others?).
Best thing i can think of, we could use an array of like 3,4 images in our builds, and try them in order from the most ancient to the newest, like "current_image,image_w_gcc9,image_w_gcc10,image_w_gcc11". 
Adding newer gcc to current image is not simple because the required deps (like binutils > 2.35 for gcc10) won't be easily satisfied.

leodido commented on cmd/root_options.go:
```diff
@@ -26,6 +26,7 @@ type RootOptions struct {
 	KernelRelease    string `validate:"required,ascii" name:"kernel release"`
 	Target           string `validate:"required,target" name:"target"`
 	KernelConfigData string `validate:"omitempty,base64" name:"kernel config data"` // fixme > tag "name" does not seem to work when used at struct level, but works when used at inner level
+	BuilderImage     string `name:"builder image"`
```
Would be cool to have a validator for it.

Most importantly, I think the default should be set here.

```suggestion
	BuilderImage     string `default:"falcosecurity/driverkit-builder:latest" name:"builder image"`
```

Feel free to pick my suggestion.

LucaGuerra replied:
I believe I wanted to do that but, please correct me if I'm wrong, when a new version of driverkit is released that string will be replaced with something that is not `falcosecurity/driverkit-builder:latest` but an image that has been generated at compile time for that specific version. I recall that by looking at the commit in the Makefile and in `buildprocessor.go`.
This is why the default we want is printed by looking at a variable in the appropriate module.

leodido replied:
https://github.com/falcosecurity/driverkit/blob/3ea18c6c767ac0c80102e469708102bd5bbc5e4f/cmd/root_options.go#L39 here with the `defaults` package setters.

LucaGuerra replied:
First of all, thanks a lot for validating the idea. This PR introduces code to do exactly what you wrote above in the yaml file and so I now know this is a good thing from the authors' perspective :)

Re. validator: makes sense, I can do it. 👍 
Re. "latest": when I read the driverkit source some time ago I tried to think about the reason why it was done like that, and I believe that it's designed to allow every version to keep working independently of changes in the moving "latest" tag. So if I have Driverkit v0.6.0 it will still work even if devs create a new builder for v0.7.0 which is incompatible with the previous one.
Re. injected tag: I think we want to retain this behavior. Do you have a better idea about how to do this cleanly? Thanks!

LucaGuerra replied:
Done as suggested (if I got it right)

leodido commented on cmd/root.go:
```diff
@@ -117,6 +118,7 @@ func NewRootCmd() *RootCmd {
 	flags.StringVar(&rootOpts.KernelConfigData, "kernelconfigdata", rootOpts.KernelConfigData, "base64 encoded kernel config data: in some systems it can be found under the /boot directory, in other it is gzip compressed under /proc")
 	flags.StringVar(&rootOpts.ModuleDeviceName, "moduledevicename", rootOpts.ModuleDeviceName, "kernel module device name (the default is falco, so the device will be under /dev/falco*)")
 	flags.StringVar(&rootOpts.ModuleDriverName, "moduledrivername", rootOpts.ModuleDriverName, "kernel module driver name, i.e. the name you see when you check installed modules via lsmod")
+	flags.StringVar(&rootOpts.BuilderImage, "builderimage", rootOpts.BuilderImage, fmt.Sprintf("docker image to be used to build the kernel module and eBPF probe. If not provided, the default image will be used. (default \"%v\")", driverbuilder.BuilderBaseImage))
```
```suggestion
	flags.StringVar(&rootOpts.BuilderImage, "builderimage", rootOpts.BuilderImage, "docker image to be used to build the kernel module and eBPF probe. If not provided, the default image will be used.")
```

No need to write the default manually here IF you pick up my other suggestion too.

leodido commented on pkg/driverbuilder/buildprocessor.go:
```diff
@@ -8,7 +8,7 @@ type BuildArchitecture string
 const BuildArchitectureX86_64 BuildArchitecture = "x86_64"
 
-var builderBaseImage = "falcosecurity/driverkit-builder:latest" // This is overwritten when using the Makefile to build
+var BuilderBaseImage = "falcosecurity/driverkit-builder:latest" // This is overwritten when using the Makefile to build
```
This can be probably avoided if you set the default value for the flag as suggested above.

leodido reviewed (commented):
This slipped my mind!

Reviewed and left some comments

LucaGuerra commented:
@leodido addressed your comments:
* Now the default field for the builder image uses the appropriate `SetDefaults()` function from the defaults library
* I have added a validator for the field. Note that perfect parsing of a Docker image pullstring is very hard to do, but I believe that restricting to letters+digits+separators with the appropriate separator rules is good enough to catch typos and user errors while building yamls or command lines.

leodido reviewed (approved):
LGTM

poiana commented:
LGTM label has been added.  <details>Git tree hash: 4baaaecaed921b92e8a6d8fe5736407f9743fe30</details>

leodido commented:
I see, you want to change that value (the builder image) at compile time
while releasing.

I’m not getting why that would be preferable to just change the value via a
flag/option (since we already have a fully featured CLI). Maybe I’m missing
something at the moment!

On Thu, 5 May 2022 at 15:38 Luca Guerra ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In cmd/root_options.go
> <https://github.com/falcosecurity/driverkit/pull/113#discussion_r865916076>
> :
>
> > @@ -26,6 +26,7 @@ type RootOptions struct {
>  	KernelRelease    string `validate:"required,ascii" name:"kernel release"`
>  	Target           string `validate:"required,target" name:"target"`
>  	KernelConfigData string `validate:"omitempty,base64" name:"kernel config data"` // fixme > tag "name" does not seem to work when used at struct level, but works when used at inner level
> +	BuilderImage     string `name:"builder image"`
>
> I believe I wanted to do that but, please correct me if I'm wrong, when a
> new version of driverkit is released that string will be replaced with
> something that is not falcosecurity/driverkit-builder:latest but an image
> that has been generated at compile time for that specific version. I recall
> that by looking at the commit in the Makefile and in buildprocessor.go.
> This is why the default we want is printed by looking at a variable in the
> appropriate module.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/falcosecurity/driverkit/pull/113#discussion_r865916076>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA5J4YUGGWZRTEYGQWAYLLVIPFMDANCNFSM5KU75HHQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
L.
